File: Array/C++/TwoSum.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class Solution {
// Brute Force O(n^2)
public:
    vector<int> twoSumBF(vector<int>& nums, int target) {
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[i] + nums[j] == target) { return {i, j}; }
            }
        }
        return {};
    }

// Optimal Solution O(n)
public:
    vector<int> twoSumOPT(vector<int>& nums, int target) {
        map<int, int> dict;
        
        for (int i = 0; i < nums.size(); i++) {
            int comp = target - nums[i];
            
            if (dict.count(comp)) {
                return {dict.at(comp), i};
            }
            
            dict.insert({nums[i], i});
        }
        throw invalid_argument("No Match Found");
    }
};

File: Array/Python/TwoSum.py
# Brute Force O(n^2)
def twoSumBF(nums, target):
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]

# Optimal Solution O(n)
def twoSumOPT(nums, target):
    hashTable = {}
    for i in range(len(nums)):
        if nums[i] not in hashTable:
            hashTable[target - nums[i]] = i
        else:
            return [hashTable[nums[i]], i]
        